header-to-metadata: add support for cookie to metadata

Radha13 opened the pull request:
Commit Message:
header-to-metadata filter supports adding a header's value to a metadata key which is later used for subset load balancing.

This PR adds support for extracting a specific cookie value before it's added as metadata.

Usage:

- example request headers:
```
'cookie', some=cookie; someother=cookie; X-Devhostname=docs-dev-yellow-mon'

```

- example config:
```
      request_rules:
        - cookie: "X-Devhostname"
          on_header_present:
            metadata_namespace: envoy.lb 
            key: "dev-hostname"
            type: STRING
            remove: false
```

- adds - metadata['dev-hostname'] = "docs-dev-yellow-mon"

Additional Description:
- refactor code a little
- `remove` action is not supported
- regex substitution works as it does for headers.

Risk Level: Medium
Testing: Unit tests
Docs Changes: updated
Release Notes: TODO

Co-authored-by: Ariane van der Steldt  <avandersteldt@slack-corp.com>
Signed-off-by: Radha Kumari <rkumari@slack-corp.com>

zuercher commented on api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto:
```diff
@@ -90,26 +90,33 @@ message Config {
-    string header = 1
-        [(validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}];
+    // Specifies one of the header or cookie match will be performed on the value of the header.
+    oneof header_cookie_specifier {
```
Couple things here:

1. We cannot change the numbering of existing fields or else we break wire-compatibility of existing control planes. It is ok for fields to be out-of-order with respect to their identifiers, so `cookie` can be 5.
2. We also cannot promote fields to be part of a `oneof`. So the way this can work for v3 is to remove the oneof and tag the two fields with `[(udpa.annotations.field_migrate).oneof_promotion = "header_cookie_specifier"]` and remove the `min_bytes: 1` validateion. Then in the code handles the Rule, we'll have to manually enforce that exactly one of those fields is non-empty. In v4 this will become a oneof automatically and when we convert Envoy to use the v4 protos we can reinstate the validation and remove validation code from the Rule handling.

zuercher commented on docs/root/configuration/http/http_filters/header_to_metadata_filter.rst:
```diff
@@ -39,6 +40,28 @@ absence of a version header could be:
               type: STRING
             remove: false
 
+Same way as header, the value of the specified cookie will be extracted from the request
```
nit: s/Same way as header/As with headers/

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,72 @@ namespace Extensions {
+absl::optional<std::string> HeaderValueSelector::extract(Http::HeaderMap& map) const {
+  const Http::HeaderEntry* header_entry = map.get(header_);
+  if (header_entry == nullptr) {
+    return absl::optional<std::string>(); // no value
```
nit: `return absl::nullopt;`

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,72 @@ namespace Extensions {
+absl::optional<std::string> CookieValueSelector::extract(Http::HeaderMap& map) const {
+  std::string value = Envoy::Http::Utility::parseCookieValue(map, cookie_);
+  if (!value.empty()) {
+    return absl::optional<std::string>(value);
```
nit: absl::nullopt

Radha13 replied:
I think you mean this for the line 41.

zuercher replied:
Well, definitely not this line. Sorry.

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,72 @@ namespace Extensions {
+  const auto& option = rule.header_cookie_specifier_case();
+
+  // initialize the shared pointer
+  switch(option) {
```
As mentioned above, we'll need to just directly check `rule.header()` and `rule.cookie()` for non-empty values, with a TODO similar to the one below re: v4.

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -48,27 +113,9 @@ bool Config::configToVector(const ProtobufRepeatedRule& proto_rules,
     return false;
   }
 
+  // either a cookie or a header?
```
nit: I don't think this comment is necessary.

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -172,56 +219,47 @@ const std::string& HeaderToMetadataFilter::decideNamespace(const std::string& ns
-      if (proto_rule.remove()) {
-        headers.remove(header);
-      }
+    bool has_value;
```
nit: just `value` and `has_value` inline and map `has_value` const. However, you can also just use `value.has_value()` or even just `value` (which will be true if set).

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.h:
```diff
@@ -20,17 +20,52 @@ namespace HeaderToMetadataFilter {
   const Regex::CompiledMatcherPtr& regexRewrite() const { return regex_rewrite_; }
   const std::string& regexSubstitution() const { return regex_rewrite_substitution_; }
-  const Http::LowerCaseString& header() const { return header_; }
+  std::shared_ptr<const ValueSelector> vs;
```
Envoy style it to append `_` to field names. We'd prefer a more descriptive name as well.

zuercher commented on test/extensions/filters/http/header_to_metadata/config_test.cc:
```diff
@@ -57,7 +57,33 @@ TEST(HeaderToMetadataFilterConfigTest, InvalidEmptyKey) {
+TEST(HeaderToMetadataFilterConfigTest, InvalidEmptyCookieHeader) {
+  const std::string yaml = R"EOF(
+request_rules:
+- header: ""
```
Should this be `- cookie: ""`?

Radha13 replied:
woops :| yeah.

zuercher commented on test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:
```diff
@@ -443,6 +443,28 @@ TEST_F(HeaderToMetadataTest, PerRouteEmtpyRules) {
+    remove: true
+)EOF";
+  auto expected = "Cannot specify remove for cookie";
+  EXPECT_THROW_WITH_MESSAGE(initializeFilter(config), EnvoyException, expected);
```
Let's document this limitation in the proto.

zuercher commented on test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:
```diff
@@ -443,6 +443,28 @@ TEST_F(HeaderToMetadataTest, PerRouteEmtpyRules) {
+
+TEST_F(HeaderToMetadataTest, RejectRemoveCookieRules) {
+  envoy::extensions::filters::http::header_to_metadata::v3::Config config_proto;
+  EXPECT_THROW(std::make_shared<Config>(config_proto, true), EnvoyException);
```
I think this is a duplicate of the PerRouteEmtpyRules test, and can be removed.

zuercher commented on api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto:
```diff
@@ -90,9 +90,19 @@ message Config {
-    // The header that triggers this rule — required.
-    string header = 1
-        [(validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}];
+    // Specifies one of the header or cookie match will be performed on the value of the header.
```
nit: "Specifies that"

zuercher commented on test/extensions/filters/http/header_to_metadata/config_test.cc:
```diff
@@ -31,22 +31,26 @@ void testForbiddenConfig(const std::string& yaml) {
-// Tests that an empty header is rejected.
-TEST(HeaderToMetadataFilterConfigTest, InvalidEmptyHeader) {
+// Tests that empty (metadata) keys are rejected.
+TEST(HeaderToMetadataFilterConfigTest, InvalidEmptyKey) {
```
Also, I think we're missing a test that attempts to configure a rule with both `header` and `cookie` set.

Radha13 replied:
One thing that I have been fuzzy about it, where should the test live, config or filter?

rgs1 replied:
technically config -- since it should be rejected at config time. but there are some tests that exercise config that are in the filter test file... we can clean it up later tho.

Radha13 replied:
Added both the tests below.

zuercher commented:
Oops. Forgot to mention that you'll need re-apply some of those recent commits with the DCO signoff. There's some instructions on this in the CONTRIBUTING.md doc, but basically amend the commits with the signoff line and force-push.

@envoyproxy/api-shepherds need to have a look too.

rgs1 commented on docs/root/configuration/http/http_filters/header_to_metadata_filter.rst:
```diff
@@ -6,8 +6,9 @@ Envoy Header-To-Metadata Filter
 This filter is configured with rules that will be matched against requests and responses.
-Each rule has a header and can be triggered either when the header is present or missing. When
-a rule is triggered, dynamic metadata will be added based on the configuration of the rule.
+Each rule has either a cookie or a header and can be triggered either when the header
```
what about if the cookie is missing?

Radha13 replied:
I will document that case :) thanks.

Radha13 replied:
updated below.

rgs1 commented on docs/root/configuration/http/http_filters/header_to_metadata_filter.rst:
```diff
@@ -39,6 +40,28 @@ absence of a version header could be:
             remove: false
 
+As with headers, the value of the specified cookie will be extracted from the request
+and added as metdata with the key specified.
```
typo: metadata

rgs1 commented on docs/root/configuration/http/http_filters/header_to_metadata_filter.rst:
```diff
@@ -39,6 +40,28 @@ absence of a version header could be:
+              key: default
+              value: 'true'
+              type: STRING
+            remove: false
```
hmm it feels weird to use `on_header_present` and on `on_header_missing` to act on a cookie... Should we use `(udpa.annotations.field_migrate).rename = "on_present"` (and `on_missing`) to have these renamed in v4?

rgs1 replied:
cc: @zuercher

Radha13 replied:
I think it makes sense to have those 2 fields renamed. let me know what you both prefer, happy to make that change :)

rgs1 replied:
Yeah I like the idea of the renaming for v4. But we can do it in a follow-up PR too, to avoid delaying this one.

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,69 @@ namespace Extensions {
 namespace HeaderToMetadataFilter {
 
-Rule::Rule(const std::string& header, const ProtoRule& rule) : header_(header), rule_(rule) {
+// extract the value of the header
```
nit: i think most of the other comments are full sentences (e.g.: upper case at the start, period at the end).

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,69 @@ namespace Extensions {
+  if (!value.empty()) {
+    return absl::optional<std::string>(value);
+  }
+  return absl::nullopt; // no value
```
Hmm do we want to support removing the cookie (as we do for header)?

rgs1 replied:
If we don't -- we need to call it out since there's nothing in the proto definition that says remove doesn't work for cookies.

rgs1 replied:
nvm -- just saw there is a comment. we are good.

rgs1 replied:
nit: drop the comment too

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,69 @@ namespace Extensions {
+    selector_ = std::make_shared<CookieValueSelector>(rule.cookie());
+  } else {
+    throw EnvoyException("One of Cookie or Header option needs to be specified");
+  }
```
this else block is redundant.... might as well remove it.

Radha13 replied:
I'm not sure. we used to have a test that checked whether a header was specified (I removed it but I plan to add it back). If we remove this the test will fail. Which means that we are not ensuring either one of header or cookie must be specified.

rgs1 replied:
Ah -- I see. Ok.

rgs1 commented on api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto:
```diff
@@ -90,9 +90,19 @@ message Config {
+       (udpa.annotations.field_migrate).oneof_promotion = "header_cookie_specifier"
+    ];
 
     // If the header is present, apply this metadata KeyValuePair.
```
If the header or cookie....

rgs1 replied:
ditto for `on_header_missing`

rgs1 commented:
Looks good, just a few comments here and there. Thanks!

Radha13 commented:
Hi everyone. Really appreciate all the feedback.
It seems that I messed up the commit sign off real bad and since I had to merge from master twice, the provided DCO steps are not working for me and might have to squash the commits.
My apologies in advance.

rgs1 commented on docs/root/configuration/http/http_filters/header_to_metadata_filter.rst:
```diff
@@ -39,6 +45,29 @@ absence of a version header could be:
 
+As with headers, the value of the specified cookie will be extracted from the request
+and added as metadata with the key specified.
+`remove: true` is unsupported for cookie.
```
nit: maybe "Removing a cookie when a rule matches is unsupported."?

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,70 @@ namespace Extensions {
+absl::optional<std::string> HeaderValueSelector::extract(Http::HeaderMap& map) const {
+  const Http::HeaderEntry* header_entry = map.get(header_);
+  if (header_entry == nullptr) {
+    return absl::nullopt; // no value
```
nit: drop the comment (i think the method is already nice & easy to follow).

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -18,7 +18,70 @@ namespace Extensions {
+  // Catch the value in the header before removing.
+  absl::optional<std::string> value = std::string(header_entry->value().getStringView());
+  if (remove_) {
+    map.remove(header_); // remove the header
```
ditto

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -120,7 +164,7 @@ bool HeaderToMetadataFilter::addMetadata(StructMap& map, const std::string& meta
   }
 
-  std::string decodedValue = std::string(value);
+  std::string decodedValue = value;
```
Now that `value` is an std::string that this method owns we can just use that directly, without copying it into `decodedValue` first.

rgs1 commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -120,7 +164,7 @@ bool HeaderToMetadataFilter::addMetadata(StructMap& map, const std::string& meta
-  std::string decodedValue = std::string(value);
+  std::string decodedValue = value;
   if (encode == envoy::extensions::filters::http::header_to_metadata::v3::Config::BASE64) {
     decodedValue = Base64::decodeWithoutPadding(value);
```
drop the above and make this:

```
value = Base64::decodeWithoutPadding(value);
```

and just use `value` for the rest of the method.

rgs1 commented on test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:
```diff
@@ -443,6 +443,55 @@ TEST_F(HeaderToMetadataTest, PerRouteEmtpyRules) {
 
+/**
+ * Invalid empty header or cookie should be rejected.
+ */
```
nit:

```
// Invalid empty header or cookie should be rejected.
```

Radha13 replied:
@rgs1 the only reason I used multi-line comment is because every other test in that file had that and I wanted to keep it consistent. should I change all of them?

rgs1 replied:
Ah gotcha, it's good then.

rgs1 commented on test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:
```diff
@@ -443,6 +443,55 @@ TEST_F(HeaderToMetadataTest, PerRouteEmtpyRules) {
+
+/**
+ * Rules with both header and cookie fields should be rejected.
+ */
```
nit:

```
// Rules with both header and cookie fields should be rejected.
```

rgs1 commented on test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:
```diff
@@ -443,6 +443,55 @@ TEST_F(HeaderToMetadataTest, PerRouteEmtpyRules) {
+
+/**
+ * Rules with remove field should be rejected in case of a cookie.
+ */
```
ditto

rgs1 commented on test/extensions/filters/http/header_to_metadata/header_to_metadata_filter_test.cc:
```diff
@@ -543,6 +592,163 @@ TEST_F(HeaderToMetadataTest, NoMissingWhenHeaderIsPresent) {
+/**
+ * on header missing case with no header data
+ */
+
```
ditto

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.h:
```diff
@@ -20,17 +20,52 @@ namespace HeaderToMetadataFilter {
+// Interface for getting values.
+class ValueSelector {
+public:
+  virtual absl::optional<std::string> extract(Http::HeaderMap& map) const PURE;
```
Lets add `@return absl::optional<std::string> the extracted header or cookie`

Radha13 commented:
So the DCO is fixed and I have addressed most of the review comment (waiting on some clarification on the last one.) Please let me know if there is some thing else that I missed.

Radha13 commented:
Thank you so much @zuercher and @rgs1 I think I have addressed all the review comments :)

zuercher commented on api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto:
```diff
@@ -90,25 +90,36 @@ message Config {
-        [(validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}];
-
-    // If the header is present, apply this metadata KeyValuePair.
+    // Specifies that a match will be performed on the value of a header or a cookie.
+    //
+    // The header to be extracted.
```
To make the code formatter happier.

```suggestion
    // Specifies that a match will be performed on the value of a header.
```

zuercher commented on api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto:
```diff
@@ -90,25 +90,36 @@ message Config {
+      (udpa.annotations.field_migrate).oneof_promotion = "header_cookie_specifier"
+    ];
+
+    // The cookie to be extracted.
```
```suggestion
    // Specifies that a match will be performed on the value of a cookie.
```

zuercher commented on api/envoy/extensions/filters/http/header_to_metadata/v4alpha/header_to_metadata.proto:
```diff
@@ -90,25 +90,35 @@ message Config {
-    // The header that triggers this rule — required.
-    string header = 1
-        [(validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}];
+    // Specifies that a match will be performed on the value of a header or a cookie.
```
More formatter: Delete this, and update the comments for `header` and `cookie` to match the v3 comments.

Radha13 replied:
the checker wants me to remove `min_bytes:1` from there as well

zuercher commented on api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto:
```diff
@@ -90,25 +90,36 @@ message Config {
     option (udpa.annotations.versioning).previous_message_type =
```
Add `// [#next-free-field: 6]` as a single line between the comment and `message Rule` (and same for the v4alpha version).

Radha13 commented:
So looked at https://dev.azure.com/cncf/envoy/_build/results?buildId=46226&view=logs&j=9fd74226-df5b-5f29-39da-76fc341cc763&t=068b98b1-acf0-52d0-fc6e-f6d1a5b1d5ef&l=6812 and think I got most but will see why the checks/test runs again
